### PR TITLE
Enabled DETR (Object Detection) model

### PR DESCRIPTION
**Pull request by cfgfung**
# What does this PR do?

This PR contains the patch, example, and test codes for DETR models.

A100 CUDA BF16(Autocast) benchmarks:
n_iterations: 10
Total latency (ms): 241.29199981689453
Average latency (ms): 24.129199981689453

Gaudi2 BF16(Autocast and Graph mode) benchmarks: 
n_iterations: 10
Total latency (ms): 65.3073787689209
Average latency (ms): 6.53073787689209

## Before submitting
- [N.A.] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [Yes] Did you make sure to update the documentation with your changes?
- [Yes] Did you write any new necessary tests?

### CODE REVIEW

**Comment by libinta** [comment]
@cfgfung  can you rebase the PR?

**Comment by cfgfung** [comment]
> @cfgfung can you rebase the PR?

Hi,
I have applied rebase and here is the test result

image

**Review by imangohari1 — CHANGES_REQUESTED** [review]
Hi @cfgfung 
thank you for the work here. 
Below are my suggestions for this PR. Let's work on these and do a final review. 
 
 - I've made some changes regarding some minor clean ups and ci tests. Please review and implement them via `git am < 0001*`  (file attached). 
 - Few minor comments are give in comment. 

[0001-fea-Minor-CI-updates-and-clean-ups.patch](https://github.com/user-attachments/files/16017846/0001-fea-Minor-CI-updates-and-clean-ups.patch)

File: Makefile [hunk 1]
```diff
@@ -41,6 +41,12 @@ fast_tests_diffusers:
 	python -m pip install .[tests]
 	python -m pytest tests/test_diffusers.py
 
+# Run unit and integration tests related to Image segmentation
+fast_tests_object_detection:
+	python -m pip install .[tests]
+	python -m pip install timm
```
Comment (imangohari1) [inline Makefile:47]: if no_timm is used for the test, is this needed?

File: examples/object-detection/run_example.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    parser.add_argument("--warmup", type=int, default=3, help="Number of warmup iterations for benchmarking.")
+    parser.add_argument(
+        "--n_iterations", type=int, default=10, help="Number of inference iterations for benchmarking."
+    )
+
+    args = parser.parse_args()
+
+    adapt_transformers_to_gaudi()
+
+    # you can specify the revision tag if you don't want the timm dependency
```
Comment (imangohari1) [inline examples/object-detection/run_example.py:70]: not sure how to approach this here, but would be nice to able to pass/toggle between `no_timm` and `main` revisions.

**Review by imangohari1 — COMMENTED** [review]
minor comments.

**Comment by imangohari1** [comment]
@cfgfung 
Can we move this pr along? Pls. apply the changes I suggested. Tnx.

File: Makefile (hunk 1 above)
Comment (cfgfung) [reply Makefile:47]: Yes, it is still needed. The model has called some API/modules from timm library

**Comment by cfgfung** [comment]
> Hi @cfgfung thank you for the work here. Below are my suggestions for this PR. Let's work on these and do a final review.
> 
> * I've made some changes regarding some minor clean ups and ci tests. Please review and implement them via `git am < 0001*`  (file attached).
> * Few minor comments are give in comment.
> 
> [0001-fea-Minor-CI-updates-and-clean-ups.patch](https://github.com/user-attachments/files/16017846/0001-fea-Minor-CI-updates-and-clean-ups.patch)
> 
> G2 results (after applying the patch)
> 
> ```shell
> ---------------------------: System Configuration :---------------------------
> Num CPU Cores : 160
> CPU RAM       : 1056375224 KB
> ------------------------------------------------------------------------------
> Detected cat with confidence 1.0 at location [344.0, 25.25, 640.0, 376.0]
> Detected remote with confidence 0.996 at location [328.0, 75.5, 372.0, 188.0]
> Detected remote with confidence 0.996 at location [39.0, 70.5, 176.0, 118.0]
> Detected cat with confidence 1.0 at location [15.62, 52.5, 316.0, 472.0]
> Detected couch with confidence 0.996 at location [-1.25, 0.94, 640.0, 472.0]
> 
> Stats:
> ------------------------------------------------------------
> Total latency (ms): 58.38346481323242 (for n_iterations=10)
> Average latency (ms): 5.838346481323242 (per iteration)
> ------------------------------------------------------------
> ```

I have applied the patch.

File: examples/object-detection/run_example.py (hunk 2 above)
Comment (cfgfung) [reply examples/object-detection/run_example.py:70]: Hi,  
The DETR is composed of different modules from timm, transformer and other popular libraries. 
Calling this line will apply any applicable optimizations to this DETR model, especially the transformer submodules.

**Comment by cfgfung** [comment]
@imangohari1 Hi, I have applied rebase and addressed the changes mentioned above.

File: Makefile (hunk 1 above)
Comment (imangohari1) [reply Makefile:67]: then let's move this to here: https://github.com/huggingface/optimum-habana/blob/main/setup.py#L41-L51
doesn't need to be a separate pip install.

**Review by imangohari1 — CHANGES_REQUESTED** [review]
minor change. 
Also could we test this with multiple input images?

Please make sure to run `make style` and rebase/sync with the head of OH at main.

**Comment by cfgfung** [comment]
> minor change. Also could we test this with multiple input images?
> 
> Please make sure to run `make style` and rebase/sync with the head of OH at main.

Moved the timm installation command to setup.py. This is the result for the `make style`
image

A single image should be enough to safeguard the precision/numerical issues and throughput. The following merged PRs are also using a single image as a test.

https://github.com/huggingface/optimum-habana/pull/801
https://github.com/huggingface/optimum-habana/pull/814
https://github.com/huggingface/optimum-habana/pull/826

**Review by imangohari1 — APPROVED** [review]
Looks good. 

@regisss 
Could you take a final look here please?
Thanks

File: setup.py [hunk 3]
```diff
@@ -48,6 +48,7 @@
     "safetensors",
     "pytest < 8.0.0",
     "torchsde",
+    "timm",
```
Comment (regisss) [inline setup.py:51]: Can you tell a bit more about why timm is needed?

Comment (regisss) [reply setup.py:51]: Okay, I saw the discussion above

Comment (regisss) [reply setup.py:51]: But I guess it's needed for the example too no?

Comment (cfgfung) [reply setup.py:51]: For the example code, I explicitly set it not to use timm module (i.e. no_timm). So we do not need to import timm for that.
`model = DetrForObjectDetection.from_pretrained("facebook/detr-resnet-101", revision="no_timm")`

**Comment by yeonsily** [comment]
@cfgfung  can you please fix code style?

**Comment by cfgfung** [comment]
> @cfgfung can you please fix code style?

Hi,

On my side I cannot see anything changed using `make style`

image

**Review by regisss — APPROVED** [review]
LGTM!

**Pull request merged**